src/record.rs

```diff
@@ -99,13 +121,95 @@ impl<'a> Serializable<'a> for MultiPlexedRecord<'a> {
                 position,
                 payload,
             }),
+            RecordType::AppendRecords => Some(MultiPlexedRecord::AppendRecords {
+                queue,
+                position,
+                records: MultiRecord::new(payload)?,
+            }),
             RecordType::Truncate => Some(MultiPlexedRecord::Truncate { queue, position }),
             RecordType::Touch => Some(MultiPlexedRecord::Touch { queue, position }),
             RecordType::DeleteQueue => Some(MultiPlexedRecord::DeleteQueue { queue, position }),
         }
     }
 }
 
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct MultiRecord<'a> {
+    buffer: &'a [u8],
+    position: usize,
+}
+
+impl<'a> MultiRecord<'a> {
+    pub fn new(buffer: &[u8]) -> Option<MultiRecord> {
+        let mut mrecord = MultiRecord {
+            buffer,
+            position: 0,
+        };
+        // verify the content is not corrupted
+        for record in mrecord {
+            record?;
+        }
+
+        mrecord.reset_position();
+
+        Some(mrecord)
+    }
+
+    pub fn new_unchecked(buffer: &[u8]) -> MultiRecord {
+        MultiRecord {
+            buffer,
+            position: 0,
+        }
+    }
+
+    pub fn serialize<'b, T: Iterator<Item = &'b [u8]>>(items: T, mut position: u64) -> Vec<u8> {
+        let mut res = Vec::new();
+        for item in items {
+            res.extend_from_slice(&position.to_le_bytes());
+            position += 1;
+            res.extend_from_slice(&item.len().to_le_bytes());
+            res.extend_from_slice(item);
+        }
+        res
+    }
+
+    pub fn reset_position(&mut self) {
+        self.position = 0;
+    }
+}
+
+impl<'a> Iterator for MultiRecord<'a> {
+    type Item = Option<(u64, &'a [u8])>;
```

Comment:
None here means a corruption?

can we make that more explicit with an empty error struct.
The `Option<Option<>>` is confusing.
`struct MultiRecordCorruption;`

Can we make that a Result<(), Corruption>

Reply:
It might be better to update the rturn type of `deserialize` too